File: master_wallet/ios/TigerMasterWallet/Sources/Services/PrivacyService.swift
/**
 * PrivacyService - iOS Implementation
 *
 * On-device privacy primitives that can be done honestly are kept:
 *   - AES-256-GCM encryption/decryption of sensitive data (CryptoKit, Keychain key).
 *
 * Primitives that cannot be done client-side with real cryptography are
 * fail-closed (they throw) rather than simulated:
 *   - ZK-SNARK proof generation/verification (no on-device Groth16/PLONK
 *     prover/verifier).
 *   - CoinJoin mixing (requires a coordinator service + real range proofs).
 *   - Stealth-address derivation (requires secp256k1 + keccak256, not P256).
 *   - Address rotation (a real Ethereum address requires keccak256 of a
 *     secp256k1 public key, unavailable on-device).
 */

import Foundation
import Security
import CryptoKit
import LocalAuthentication

public enum PrivacyError: Error, LocalizedError {
    case zkProvingUnsupported
    case zkVerificationUnsupported
    case coinJoinUnsupported
    case stealthAddressUnsupported
    case addressRotationUnsupported

    public var errorDescription: String? {
        switch self {
        case .zkProvingUnsupported:
            return "ZK proof generation is not supported on-device (no Groth16/PLONK prover). Use the backend."
        case .zkVerificationUnsupported:
            return "ZK proof verification is not supported on-device (no Groth16/PLONK verifier). Use the backend."
        case .coinJoinUnsupported:
            return "CoinJoin mixing requires a coordinator service and real range proofs; not supported on-device."
        case .stealthAddressUnsupported:
            return "Stealth-address derivation requires secp256k1 + keccak256 (unavailable on-device); use the backend."
        case .addressRotationUnsupported:
            return "Real address rotation requires keccak256 of a secp256k1 public key (unavailable on-device); use the backend."
        }
    }
}

public class PrivacyService {

    // MARK: - Singleton
    public static let shared = PrivacyService()

    // Privacy levels (kept for API compatibility with callers that reference them)
    public static let PRIVACY_NONE = 0
    public static let PRIVACY_STANDARD = 1
    public static let PRIVACY_HIGH = 2
    public static let PRIVACY_MAXIMUM = 3

    // MARK: - Initialization
    private init() {}

    // MARK: - Stealth Address (fail-closed)

    public func generateStealthAddress(ownerAddress: String, spendingPublicKey: Data,
                                       completion: @escaping (StealthAddressResult) -> Void) {
        completion(StealthAddressResult(success: false, error: PrivacyError.stealthAddressUnsupported.localizedDescription))
    }

    // MARK: - CoinJoin (fail-closed)

    public func createCoinJoin(inputs: [CoinJoinInput], outputs: [CoinJoinOutput], privacyLevel: Int,
                                completion: @escaping (CoinJoinResult) -> Void) {
        completion(CoinJoinResult(success: false, error: PrivacyError.coinJoinUnsupported.localizedDescription))
    }

    // MARK: - ZK Proof (fail-closed)

    public func generateZKProof(amount: String, commitment: Data,
                                completion: @escaping (ZKProofResult) -> Void) {
        completion(ZKProofResult(success: false, error: PrivacyError.zkProvingUnsupported.localizedDescription))
    }

    /// Verify ZK proof. There is no on-device Groth16/PLONK verifier, so this
    /// always fails closed. An empty/all-zero proof is explicitly rejected
    /// before throwing so a caller cannot treat absence as success.
    public func verifyZKProof(proof: String, commitment: Data) -> Bool {
        let trimmed = proof.trimmingCharacters(in: .whitespaces)
        let isAllZero = trimmed.allSatisfy { $0 == "0" || $0 == "x" || $0 == "X" }
        if trimmed.isEmpty || (commitment.isEmpty) {
            return false
        }
        if isAllZero {
            return false
        }
        return false
    }

    // MARK: - Address Rotation (fail-closed)

    public func rotateAddress(currentAddress: String,
                              completion: @escaping (RotationResult) -> Void) {
        completion(RotationResult(success: false, error: PrivacyError.addressRotationUnsupported.localizedDescription))
    }

    // MARK: - Encryption (real CryptoKit AES-256-GCM)

    /// Encrypt sensitive data with a hardware-backed symmetric key.
    public func encryptSensitiveData(_ data: Data, completion: @escaping (EncryptedDataResult) -> Void) {
        DispatchQueue.global(qos: .userInitiated).async {
            do {
                let key = try self.getOrCreatePrivacyKey()
                let sealedBox = try AES.GCM.seal(data, using: key)
                let combined = sealedBox.nonce + sealedBox.ciphertext + sealedBox.tag

                completion(EncryptedDataResult(
                    success: true,
                    encryptedData: combined.base64EncodedString()
                ))
            } catch {
                completion(EncryptedDataResult(success: false, error: error.localizedDescription))
            }
        }
    }

    /// Decrypt sensitive data.
    public func decryptSensitiveData(_ encryptedBase64: String, completion: @escaping (DecryptedDataResult) -> Void) {
        DispatchQueue.global(qos: .userInitiated).async {
            do {
                guard let combined = Data(base64Encoded: encryptedBase64) else {
                    completion(DecryptedDataResult(success: false, error: "Invalid data"))
                    return
                }

                let key = try self.getOrCreatePrivacyKey()
                let nonce = combined.prefix(12)
                let ciphertext = combined.dropFirst(12).dropLast(16)
                let tag = combined.suffix(16)

                let sealedBox = try AES.GCM.SealedBox(nonce: nonce, ciphertext: ciphertext, tag: tag)
                let decrypted = try AES.GCM.open(sealedBox, using: key)

                completion(DecryptedDataResult(success: true, data: decrypted))
            } catch {
                completion(DecryptedDataResult(success: false, error: error.localizedDescription))
            }
        }
    }

    // MARK: - Private Helpers

    private func getOrCreatePrivacyKey() throws -> SymmetricKey {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrAccount as String: "tigermaster_privacy_key",
            kSecReturnData as String: true
        ]

        var result: AnyObject?
        let status = SecItemCopyMatching(query as CFDictionary, &result)

        if status == errSecSuccess, let keyData = result as? Data {
            return SymmetricKey(data: keyData)
        }

        let key = SymmetricKey(size: .bits256)

        let addQuery: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrAccount as String: "tigermaster_privacy_key",
            kSecValueData as String: key.withUnsafeBytes { Data($0) },
            kSecAttrAccessible as String: kSecAttrAccessibleWhenUnlockedThisDeviceOnly
        ]

        SecItemAdd(addQuery as CFDictionary, nil)

        return key
    }
}

// MARK: - Data Structures

public struct CoinJoinInput {
    public let address: String
    public let amount: String
    public let privateKey: Data
}

public struct CoinJoinOutput {
    public let address: String
    public let amount: String
}

public struct StealthAddressResult {
    public let success: Bool
    public let stealthAddress: String
    public let viewingKey: String
    public let ephemeralPublicKey: String
    public let error: String?

    public init(success: Bool, stealthAddress: String = "", viewingKey: String = "",
                ephemeralPublicKey: String = "", error: String? = nil) {
        self.success = success
        self.stealthAddress = stealthAddress
        self.viewingKey = viewingKey
        self.ephemeralPublicKey = ephemeralPublicKey
        self.error = error
    }
}

public struct CoinJoinResult {
    public let success: Bool
    public let mixedOutputs: [String]
    public let proofs: [Data]
    public let rounds: Int
    public let error: String?

    public init(success: Bool, mixedOutputs: [String] = [], proofs: [Data] = [],
                rounds: Int = 0, error: String? = nil) {
        self.success = success
        self.mixedOutputs = mixedOutputs
        self.proofs = proofs
        self.rounds = rounds
        self.error = error
    }
}

public struct ZKProofResult {
    public let success: Bool
    public let proof: String
    public let commitment: String
    public let blindingFactor: String
    public let error: String?

    public init(success: Bool, proof: String = "", commitment: String = "",
                blindingFactor: String = "", error: String? = nil) {
        self.success = success
        self.proof = proof
        self.commitment = commitment
        self.blindingFactor = blindingFactor
        self.error = error
    }
}

public struct RotationResult {
    public let success: Bool
    public let newAddress: String
    public let newPublicKey: String
    public let viewingKey: String
    public let error: String?

    public init(success: Bool, newAddress: String = "", newPublicKey: String = "",
                viewingKey: String = "", error: String? = nil) {
        self.success = success
        self.newAddress = newAddress
        self.newPublicKey = newPublicKey
        self.viewingKey = viewingKey
        self.error = error
    }
}

public struct EncryptedDataResult {
    public let success: Bool
    public let encryptedData: String
    public let error: String?

    public init(success: Bool, encryptedData: String = "", error: String? = nil) {
        self.success = success
        self.encryptedData = encryptedData
        self.error = error
    }
}

public struct DecryptedDataResult {
    public let success: Bool
    public let data: Data
    public let error: String?

    public init(success: Bool, data: Data = Data(), error: String? = nil) {
        self.success = success
        self.data = data
        self.error = error
    }
}

File: master_wallet/android/app/src/main/java/com/tigermaster/services/PrivacyService.kt
/**
 * PrivacyService - Android Implementation
 *
 * Real, hardware-backed AES-GCM encryption of sensitive data (below) is retained.
 * The ZK-proof / stealth-address / CoinJoin / address-rotation features require a
 * proving system and secp256k1 ECDH curve math that are not available on this client
 * and have no canonical backend endpoint. They fail closed rather than fabricate
 * proofs, addresses, or signatures.
 */

package com.tigermaster.services

import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyProperties
import android.util.Base64
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.math.BigInteger
import java.security.KeyStore
import javax.crypto.Cipher
import javax.crypto.KeyGenerator
import javax.crypto.SecretKey
import javax.crypto.spec.GCMParameterSpec

class PrivacyService {
    private val keyStore: KeyStore = KeyStore.getInstance("AndroidKeyStore").apply { load(null) }

    companion object {
        private const val ANDROID_KEYSTORE = "AndroidKeyStore"
        private const val PRIVACY_KEY_ALIAS = "tigermaster_privacy_key"
        private const val TRANSFORMATION = "AES/GCM/NoPadding"
        private const val GCM_TAG_LENGTH = 128
        private const val GCM_IV_LENGTH = 12

        const val PRIVACY_NONE = 0
        const val PRIVACY_STANDARD = 1
        const val PRIVACY_HIGH = 2
        const val PRIVACY_MAXIMUM = 3
    }

    /**
     * Stealth-address generation requires real ECDH on secp256k1 + keccak256, which is
     * not available on this client and has no canonical backend endpoint. Fail closed.
     */
    suspend fun generateStealthAddress(ownerAddress: String, spendingPublicKey: ByteArray): StealthAddress = withContext(Dispatchers.IO) {
        StealthAddress(success = false, error = "Stealth-address generation is not available on this client")
    }

    /**
     * CoinJoin mixing is not supported by the canonical backend and cannot be safely
     * constructed client-side. Fail closed.
     */
    suspend fun createCoinJoin(
        inputs: List<CoinJoinInput>,
        outputs: List<CoinJoinOutput>,
        privacyLevel: Int
    ): CoinJoinResult = withContext(Dispatchers.IO) {
        CoinJoinResult(success = false, error = "CoinJoin is not supported by the canonical backend")
    }

    /**
     * Real ZK-SNARK proof generation requires a proving system and circuit artifacts
     * not present on this client. Fail closed rather than return a fake proof.
     */
    suspend fun generateZKProof(amount: BigInteger, commitment: ByteArray): ZKProofResult = withContext(Dispatchers.IO) {
        ZKProofResult(success = false, error = "ZK proof generation is not available on this client")
    }

    /**
     * Verify a ZK proof. NEVER accept a proof on a non-empty check; without a real
     * verifier (and the matching verifying key / circuit) verification cannot be
     * performed, so this fails closed.
     */
    suspend fun verifyZKProof(proof: String, commitment: ByteArray): Boolean = withContext(Dispatchers.IO) {
        false
    }

    /**
     * Address rotation requires deriving a new secp256k1 keypair, unavailable here.
     * Fail closed.
     */
    suspend fun rotateAddress(currentAddress: String): RotationResult = withContext(Dispatchers.IO) {
        RotationResult(success = false, error = "Address rotation is not available on this client")
    }

    /**
     * Encrypt sensitive data with a hardware-backed AES-GCM key (real).
     */
    suspend fun encryptSensitiveData(data: ByteArray): EncryptedData = withContext(Dispatchers.IO) {
        try {
            val key = getOrCreatePrivacyKey()
            val cipher = Cipher.getInstance(TRANSFORMATION)
            cipher.init(Cipher.ENCRYPT_MODE, key)

            val iv = cipher.iv
            val encrypted = cipher.doFinal(data)

            val combined = ByteArray(iv.size + encrypted.size)
            System.arraycopy(iv, 0, combined, 0, iv.size)
            System.arraycopy(encrypted, 0, combined, iv.size, encrypted.size)

            EncryptedData(
                success = true,
                encrypted = Base64.encodeToString(combined, Base64.NO_WRAP)
            )
        } catch (e: Exception) {
            EncryptedData(success = false, error = e.message)
        }
    }

    /**
     * Decrypt sensitive data with the hardware-backed AES-GCM key (real).
     */
    suspend fun decryptSensitiveData(encryptedBase64: String): DecryptedData = withContext(Dispatchers.IO) {
        try {
            val key = getOrCreatePrivacyKey()
            val combined = Base64.decode(encryptedBase64, Base64.NO_WRAP)

            val iv = combined.copyOfRange(0, GCM_IV_LENGTH)
            val encrypted = combined.copyOfRange(GCM_IV_LENGTH, combined.size)

            val cipher = Cipher.getInstance(TRANSFORMATION)
            val spec = GCMParameterSpec(GCM_TAG_LENGTH, iv)
            cipher.init(Cipher.DECRYPT_MODE, key, spec)

            val decrypted = cipher.doFinal(encrypted)

            DecryptedData(
                success = true,
                data = decrypted
            )
        } catch (e: Exception) {
            DecryptedData(success = false, error = e.message)
        }
    }

    private fun getOrCreatePrivacyKey(): SecretKey {
        return if (keyStore.containsAlias(PRIVACY_KEY_ALIAS)) {
            keyStore.getKey(PRIVACY_KEY_ALIAS, null) as SecretKey
        } else {
            val keyGenerator = KeyGenerator.getInstance(KeyProperties.KEY_ALGORITHM_AES, ANDROID_KEYSTORE)
            val spec = KeyGenParameterSpec.Builder(
                PRIVACY_KEY_ALIAS,
                KeyProperties.PURPOSE_ENCRYPT or KeyProperties.PURPOSE_DECRYPT
            )
                .setBlockModes(KeyProperties.BLOCK_MODE_GCM)
                .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_NONE)
                .setKeySize(256)
                .build()

            keyGenerator.init(spec)
            keyGenerator.generateKey()
        }
    }
}

// Data classes

data class KeyPairResult(
    val privateKey: ByteArray,
    val publicKey: ByteArray
)

data class StealthAddress(
    val success: Boolean,
    val stealthAddress: String = "",
    val viewingKey: String = "",
    val ephemeralPublicKey: String = "",
    val error: String? = null
)

data class CoinJoinInput(
    val address: String,
    val amount: BigInteger,
    val privateKey: ByteArray
)

data class CoinJoinOutput(
    val address: String,
    val amount: BigInteger
)

data class CoinJoinResult(
    val success: Boolean,
    val mixedOutputs: List<String> = emptyList(),
    val proofs: List<ByteArray> = emptyList(),
    val rounds: Int = 0,
    val error: String? = null
)

data class ZKProofResult(
    val success: Boolean,
    val proof: String = "",
    val commitment: String = "",
    val blindingFactor: String = "",
    val error: String? = null
)

data class RotationResult(
    val success: Boolean,
    val newAddress: String = "",
    val newPublicKey: String = "",
    val viewingKey: String = "",
    val error: String? = null
)

data class EncryptedData(
    val success: Boolean,
    val encrypted: String = "",
    val error: String? = null
)

data class DecryptedData(
    val success: Boolean,
    val data: ByteArray = ByteArray(0),
    val error: String? = null
)
